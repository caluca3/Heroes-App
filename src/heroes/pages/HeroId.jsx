
import React, { useMemo } from 'react'
import { Navigate, useNavigate, useParams } from 'react-router-dom';
import { getHeroesById } from '../helpers';
//import '../Styles/HeroCard.css';

export const HeroId = () => {

    const { id } = useParams();
    
    const hero = useMemo(() => getHeroesById(id), [id]);


    const navigate = useNavigate();

    if (!hero) {
        return <Navigate to="/dc" />;
    }

    const onNavigateBack = () => {
      navigate(-1);  

    }

    if ( !hero ) {
      return <Navigate to="/dc"/>
    }
    
    return (<>
      <div className="card">
      <div className="row mt-5">
        <div className="col-4">
          <img 
            src={ `/assets/heroes/${ id }.jpg` } 
            alt={ hero.superhero }
            className="img-thumbnail animate__animated animate__fadeInLeft"
          />
        </div>
  
        <div className="col-8">
          <h1>{ hero.superhero }</h1>
          <ul className="list-group list-group-flush">
            <li className="list-group-item"> <b>Alter ego:</b> { hero.alter_ego } </li>
            <li className="list-group-item"> <b>Publisher:</b> { hero.publisher } </li>
            <li className="list-group-item"> <b>First appearance:</b> { hero.first_appearance } </li>
          </ul>
  
          <h5 className="mt-3"> Characters </h5>
          <p>{ hero.characters }</p>
  
          <button 
            className="btn btn-outline-warning mt-3"
            onClick={ onNavigateBack }
          >
           ← Back
          </button>
  
        </div>
      </div>
      </div>
    </>
    ); 
  }

