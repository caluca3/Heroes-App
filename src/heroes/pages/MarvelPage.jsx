import { HeroLista } from '../components';

export const MarvelPage = ()=> {
  
  return (
    <>
    <h1>Pagin marvel</h1>
    <hr/>
    <HeroLista publisher='Marvel Comics'/>
    
      
    </>
  )
}
