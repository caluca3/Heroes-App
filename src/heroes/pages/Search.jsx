import React from 'react';
import { useLocation, useNavigate } from 'react-router-dom';
import queryString from 'query-string';
import { useForm } from '../../auth/hooks/useForm';
import { HeroCard } from '../components/HeroCard';
import { getHeoresByName } from '../helpers/getHeroesByName';

export const Search = () => {

const navigate = useNavigate()
const location = useLocation()

const {q=''} = queryString.parse(location.search);
const heroes = getHeoresByName(q);

const {searchText, onInputChange} = useForm({
    searchText:q
  });

const showSearch = (q.length ===0);
const showError = (q.length >0) && heroes.length === 0;


  const onSearchSubmit =(e)=>{
    e.preventDefault();

    if (searchText.trim().length <=1) return;

    navigate(`?q=${searchText}`);

  }

  return (
    <>
          <h1>Search</h1>
          <br/>
          <div className="row">
            <div className='col-5'>
                <h3>Searching</h3>
                <hr />
                <form onSubmit={onSearchSubmit} aria-label='form'> 
                  <input 
                    type="text"
                    placeholder='Search a Hero'
                    className='form-control'
                    name='searchText'
                    autoComplete='off'
                    value={searchText}
                    onChange={onInputChange}
                   />
                  <button className='btn btn-outline-primary mt-3'>
                    Search Heroe
                  </button>
                </form>
            </div>
            <div className='col-7'>
                <h3>Result</h3>
                <hr />

                <div aria-label='alert-danger' className='alert alert-primary' 
                style={{display: showSearch?'':'none'}}>
                 <p>Searh a Hereo</p> 
                </div>

                <div aria-label='alert-primary' className='alert alert-danger' style={{display: showError ? '':'none'}}>
                 <b>No results found {q}</b> 
                </div>
                  {
                    heroes.map(hero=>(
                      <HeroCard key={hero.id}{...hero}/>
                    ))
                  }
            </div>
          </div>          
    </>
  )
}
