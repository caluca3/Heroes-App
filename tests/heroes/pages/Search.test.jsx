import { render,screen,fireEvent } from "@testing-library/react";
import { MemoryRouter } from "react-router-dom";
import { Search } from "../../../src/heroes/pages/Search";
const mockUseNavigate = jest.fn();

jest.mock('react-router-dom',()=>({
    ...jest.requireActual('react-router-dom'),
     useNavigate:()=> mockUseNavigate
     }));
     
describe('Pruebas en <Search/>', () => { 

    beforeEach(()=>{
        jest.clearAllMocks();
    })
    
    test('Debe mostrar correctamente', () => { 
        
            const {container}=render(
                <MemoryRouter>
                    <Search/>
                </MemoryRouter>
            )
            // screen.debug();
            expect(container).toMatchSnapshot();
     });
     test('Muestra el ?q', () => { 
        
        render(<MemoryRouter initialEntries={["/search?q=batman"]}>
                <Search/>
            </MemoryRouter>
        )
            const input = screen.getByRole('textbox');
            expect(input.value).toBe('batman')

            const img = screen.getByRole('img');
            expect(img.getAttribute('src')).toBe('/assets/heroes/dc-batman.jpg');
            
            const div = screen.getByLabelText('alert-danger');
            expect(div.style.display).toBe('none');
        ;})
        
        test('Debe mostrar un error si no se encuentra el heroe (batman123)', () => {
            render(<MemoryRouter initialEntries={["/search?q=batman123"]}>
                    <Search/>
                </MemoryRouter>
            )
            const input = screen.getByRole('textbox');
            expect(input.value).toBe('batman123');

            const div = screen.getByLabelText('alert-primary');
            expect(div.style.display).toBe('');
            screen.debug();
         });

         test('Debe llamar el navigate a la pantalla nueva', () => {
            const valueInput = "superman"
            render(<MemoryRouter  initialEntries={["/search"]}>
                    <Search/>
                </MemoryRouter>
             )
            const input = screen.getByRole('textbox');
            fireEvent.change(input,{target:{name:"searchText",value: valueInput}});
            fireEvent.submit(input);
            //const form=screen.getByRole('form');
            //fireEvent.submit(form);

            expect(mockUseNavigate).toHaveBeenCalledWith(`?q=${valueInput}`); 
            screen.debug();
        });

 });