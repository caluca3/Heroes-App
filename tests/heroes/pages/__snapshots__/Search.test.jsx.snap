// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <Search/> Debe mostrar correctamente 1`] = `
<div>
  <h1>
    Search
  </h1>
  <br />
  <div
    class="row"
  >
    <div
      class="col-5"
    >
      <h3>
        Searching
      </h3>
      <hr />
      <form
        aria-label="form"
      >
        <input
          autocomplete="off"
          class="form-control"
          name="searchText"
          placeholder="Search a Hero"
          type="text"
          value=""
        />
        <button
          class="btn btn-outline-primary mt-3"
        >
          Search Heroe
        </button>
      </form>
    </div>
    <div
      class="col-7"
    >
      <h3>
        Result
      </h3>
      <hr />
      <div
        aria-label="alert-danger"
        class="alert alert-primary"
      >
        <p>
          Searh a Hereo
        </p>
      </div>
      <div
        aria-label="alert-primary"
        class="alert alert-danger"
        style="display: none;"
      >
        <b>
          No results found 
        </b>
      </div>
    </div>
  </div>
</div>
`;
